File: 백준/Gold/1987. 알파벳/알파벳.py
import sys

input = sys.stdin.readline


def DFS(graph, x, y, visited, depth):
    global answer
    answer = max(answer, depth)
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
    char = graph[x][y]
    visited[ord(char) - 65] = True

    for dx, dy in directions:
        nx, ny = x + dx, y + dy
        if 0 <= nx < R and 0 <= ny < C:
            char = graph[nx][ny]
            if not visited[ord(char) - 65]:
                DFS(graph, nx, ny, visited, depth + 1)
                visited[ord(char) - 65] = False


if __name__ == "__main__":
    R, C = map(int, input().split())

    graph = [list(input().strip()) for _ in range(R)]
    visited = [False] * 26
    answer = 0
    DFS(graph, 0, 0, visited, 1)

    print(answer)

File: 백준/Gold/1987. 알파벳/알파벳.cc
#include <iostream>
#include <vector>
#include <algorithm>


using namespace std;

vector<vector<char>> board;
vector<bool> visited(20, false);

int R, C;
int answer = 0;


void DFS(int x, int y, int depth){
    answer = max(answer, depth);
    vector<pair<int, int>> direction = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
    visited[board[x][y] - 'A'] = true;

    for (auto dir : direction){
        int nx = x + dir.first;
        int ny = y + dir.second;
        if (nx >= 0 && nx < R && ny >= 0 && ny < C){
            if (!visited[board[nx][ny] - 'A']){
                DFS(nx, ny, depth + 1);
                visited[board[nx][ny] - 'A'] = false;
            }
        }
    }

}


int main(){
    cin >> R >> C;
    board.resize(R, vector<char>(C, 0));
    for(int i=0; i<R; i++){
        for(int k=0; k<C; k++){
            cin >> board[i][k];
        }
    }

    DFS(0, 0, 1);
    cout << answer << endl;
    return 0;
}